RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,404 @@
+
+      tempVHAcc[detIdStack] = vhsInStack_Acc;
+      tempVHRej[detIdStack] = vhsInStack_Rej;
+
+      LogTrace("VectorHitBuilderAlgorithm")
+          << "For detId #" << detIdStack.rawId() << " the following VHits have been accepted:";
+      for (const auto& vhIt : vhsInStack_Acc) {
+        LogTrace("VectorHitBuilderAlgorithm") << "accepted VH: " << vhIt;
+      }
+      LogTrace("VectorHitBuilderAlgorithm")
+          << "For detId #" << detIdStack.rawId() << " the following VHits have been rejected:";
+      for (const auto& vhIt : vhsInStack_Rej) {
+        LogTrace("VectorHitBuilderAlgorithm") << "rejected VH: " << vhIt;
+      }
+    }
+  }
+
+  loadDetSetVector(tempVHAcc, vhAcc);
+  loadDetSetVector(tempVHRej, vhRej);
+
+  LogDebug("VectorHitBuilderAlgorithm") << "End run VectorHitBuilderAlgorithm ... \n";
+  return;
+}
+
+bool VectorHitBuilderAlgorithm::checkClustersCompatibilityBeforeBuilding(
+    edm::Handle<edmNew::DetSetVector<Phase2TrackerCluster1D>> clusters,
+    const detset& theLowerDetSet,
+    const detset& theUpperDetSet) {
+  if (theLowerDetSet.size() == 1 && theUpperDetSet.size() == 1)
+    return true;
+
+  //order lower clusters in u
+  std::vector<Phase2TrackerCluster1D> lowerClusters;
+  if (theLowerDetSet.size() > 1)
+    LogDebug("VectorHitBuilderAlgorithm") << " more than 1 lower cluster! " << std::endl;
+  if (theUpperDetSet.size() > 1)
+    LogDebug("VectorHitBuilderAlgorithm") << " more than 1 upper cluster! " << std::endl;
+  for (const_iterator cil = theLowerDetSet.begin(); cil != theLowerDetSet.end(); ++cil) {
+    Phase2TrackerCluster1DRef clusterLower = edmNew::makeRefTo(clusters, cil);
+    lowerClusters.push_back(*clusterLower);
+  }
+  return true;
+}
+
+bool VectorHitBuilderAlgorithm::checkClustersCompatibility(Local3DPoint& poslower,
+                                                           Local3DPoint& posupper,
+                                                           LocalError& errlower,
+                                                           LocalError& errupper) const {
+  return true;
+}
+
+//----------------------------------------------------------------------------
+//ERICA::in the DT code the global position is used to compute the alpha angle and put a cut on that.
+std::vector<std::pair<VectorHit, bool>> VectorHitBuilderAlgorithm::buildVectorHits(
+    const StackGeomDet* stack,
+    edm::Handle<edmNew::DetSetVector<Phase2TrackerCluster1D>> clusters,
+    const detset& theLowerDetSet,
+    const detset& theUpperDetSet,
+    const std::vector<bool>& phase2OTClustersToSkip) {
+  std::vector<std::pair<VectorHit, bool>> result;
+  if (checkClustersCompatibilityBeforeBuilding(clusters, theLowerDetSet, theUpperDetSet)) {
+    LogDebug("VectorHitBuilderAlgorithm") << "  compatible -> continue ... " << std::endl;
+  } else {
+    LogTrace("VectorHitBuilderAlgorithm") << "  not compatible, going to the next cluster";
+  }
+
+  std::vector<Phase2TrackerCluster1DRef> lowerClusters;
```

Comment:
do we really need alll these copying?
to sort use a heap.

Reply:
Switched to heap sort. 

If the clusters are to be sorted, I can't see a way around copying them to dedicated vectors.

Reply:
clusters shall be sorted "a priori" in the detset has done for pixel and strips

Reply:
os just remove all this here, and sort in the clusterizer

Reply:
at least for the final optimized version

Reply:
Actually clusters ARE sorted by construction. The clusterizer is linear (as the strip one).
maybe for the P layer there is a need of resorting as there are two raw, still something to be done in the clusterizer once for ever

Reply:
Ah, that's great. For now I just got rid of the sorting all together as we loop over all the clusters anyway.